\hypertarget{classAligner}{}\section{Aligner Class Reference}
\label{classAligner}\index{Aligner@{Aligner}}


A modified Smith-\/\+Waterman alignment class.  




{\ttfamily \#include $<$Aligner.\+h$>$}

\subsection*{Public Member Functions}
\begin{DoxyCompactItemize}
\item 
\hyperlink{classAligner_ae8140f4445276cb5160fffe70c86c8d9}{Aligner} (\hyperlink{structMatrix}{Matrix}$<$ double $>$, int, Gene\+\_\+class)
\item 
\mbox{\Hypertarget{classAligner_a3c12f68dc3ea2c4de4732af1f429c171}\label{classAligner_a3c12f68dc3ea2c4de4732af1f429c171}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, int, int, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a16705a7607d30d6e92452bb0aca10bc7}\label{classAligner_a16705a7607d30d6e92452bb0aca10bc7}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, int, int, std\+::set$<$ std\+::string $>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_adaec3732ded8532baed211b792c349b7}\label{classAligner_adaec3732ded8532baed211b792c349b7}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, bool, int, int, bool=false)
\item 
\mbox{\Hypertarget{classAligner_aa6423d11b9e978c7e0e93f22653678f4}\label{classAligner_aa6423d11b9e978c7e0e93f22653678f4}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, bool, int, int, std\+::set$<$ std\+::string $>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a136197d2437d17676d489f2aa7bcbb42}\label{classAligner_a136197d2437d17676d489f2aa7bcbb42}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a5f11dccc7af814ca846d4166a93014fc}\label{classAligner_a5f11dccc7af814ca846d4166a93014fc}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, std\+::set$<$ std\+::string $>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_aa7cbe2d06e0baedfb946a8f06cf57392}\label{classAligner_aa7cbe2d06e0baedfb946a8f06cf57392}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a19a8c3b8edd1f211a784bbe09012468a}\label{classAligner_a19a8c3b8edd1f211a784bbe09012468a}} 
std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ {\bfseries align\+\_\+seq} (std\+::string, double, bool, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, std\+::set$<$ std\+::string $>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_ac312522ae345f643e058e0228f4bbad5}\label{classAligner_ac312522ae345f643e058e0228f4bbad5}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries align\+\_\+seqs} (std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool)
\item 
\mbox{\Hypertarget{classAligner_a770faffd4671c854610ec53a3ed2a6d0}\label{classAligner_a770faffd4671c854610ec53a3ed2a6d0}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries align\+\_\+seqs} (std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, bool)
\item 
\mbox{\Hypertarget{classAligner_afa5cc7962e4491c89e702cdc43247c7a}\label{classAligner_afa5cc7962e4491c89e702cdc43247c7a}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries align\+\_\+seqs} (std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, int, int, bool=false)
\item 
\mbox{\Hypertarget{classAligner_ab6955eabbf093249f5466f0c936838e3}\label{classAligner_ab6955eabbf093249f5466f0c936838e3}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries align\+\_\+seqs} (std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, bool, int, int, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a9a5733b9ba9a865bbe7c49112c773ada}\label{classAligner_a9a5733b9ba9a865bbe7c49112c773ada}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries align\+\_\+seqs} (std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a8761062d3b0ac4ddad6680736ae455c7}\label{classAligner_a8761062d3b0ac4ddad6680736ae455c7}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries align\+\_\+seqs} (std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_aab5ff149de5383ad859bc0825550b05f}\label{classAligner_aab5ff149de5383ad859bc0825550b05f}} 
void {\bfseries align\+\_\+seqs} (std\+::string, std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool)
\item 
\mbox{\Hypertarget{classAligner_ad7613ed0a8bed52435978ce2fe5e4c9c}\label{classAligner_ad7613ed0a8bed52435978ce2fe5e4c9c}} 
void {\bfseries align\+\_\+seqs} (std\+::string, std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, bool)
\item 
\mbox{\Hypertarget{classAligner_abfce680922a4341648bf700ea2bf5af2}\label{classAligner_abfce680922a4341648bf700ea2bf5af2}} 
void {\bfseries align\+\_\+seqs} (std\+::string, std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, int, int, bool=false)
\item 
\mbox{\Hypertarget{classAligner_ae5a94d6655e842013cbd51b3fedf504b}\label{classAligner_ae5a94d6655e842013cbd51b3fedf504b}} 
void {\bfseries align\+\_\+seqs} (std\+::string, std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, bool, int, int, bool=false)
\item 
\mbox{\Hypertarget{classAligner_ae0a579088ee219dcbf2557861f7cc65c}\label{classAligner_ae0a579088ee219dcbf2557861f7cc65c}} 
void {\bfseries align\+\_\+seqs} (std\+::string, std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_aa85ef1888654fb387b181678207c1305}\label{classAligner_aa85ef1888654fb387b181678207c1305}} 
void {\bfseries align\+\_\+seqs} (std\+::string, std\+::vector$<$ std\+::pair$<$ const int, const std\+::string $>$$>$, double, bool, bool, std\+::unordered\+\_\+map$<$ std\+::string, std\+::pair$<$ int, int $>$$>$, bool=false)
\item 
\mbox{\Hypertarget{classAligner_a444f4cd823b3986622e3dc67aaa0f1b9}\label{classAligner_a444f4cd823b3986622e3dc67aaa0f1b9}} 
void {\bfseries write\+\_\+alignments\+\_\+seq\+\_\+csv} (std\+::string, std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$$>$)
\item 
\mbox{\Hypertarget{classAligner_a9f1f8ad19d817606f631b3774f4ebb26}\label{classAligner_a9f1f8ad19d817606f631b3774f4ebb26}} 
std\+::unordered\+\_\+map$<$ int, std\+::forward\+\_\+list$<$ \hyperlink{structAlignment__data}{Alignment\+\_\+data} $>$ $>$ {\bfseries read\+\_\+alignments\+\_\+seq\+\_\+csv} (std\+::string, double, bool)
\item 
\mbox{\Hypertarget{classAligner_ae497684fa0f8f48f973f39d13bdb32b3}\label{classAligner_ae497684fa0f8f48f973f39d13bdb32b3}} 
void {\bfseries set\+\_\+genomic\+\_\+sequences} (std\+::vector$<$ std\+::pair$<$ std\+::string, std\+::string $>$ $>$)
\item 
\mbox{\Hypertarget{classAligner_a76cc0d616dbb84b6a777d16e60036b3a}\label{classAligner_a76cc0d616dbb84b6a777d16e60036b3a}} 
int {\bfseries incorporate\+\_\+in\+\_\+dels} (std\+::string \&, std\+::string \&, const std\+::forward\+\_\+list$<$ int $>$, const std\+::forward\+\_\+list$<$ int $>$, int)
\end{DoxyCompactItemize}


\subsection{Detailed Description}
A modified Smith-\/\+Waterman alignment class. 

\begin{DoxyAuthor}{Author}
Q.\+Marcou 
\end{DoxyAuthor}
\begin{DoxyVersion}{Version}
1.\+0
\end{DoxyVersion}
The \hyperlink{classAligner}{Aligner} class allows to perform SW alignments according to the parameters (substitution matrix,gap penalty) supplied upon construction of the object. The SW alignments matrix has been altered for V and J in order to allow for deletions on the deleted side only. Alignments can be made in parallel using open\+MP 

\subsection{Constructor \& Destructor Documentation}
\mbox{\Hypertarget{classAligner_ae8140f4445276cb5160fffe70c86c8d9}\label{classAligner_ae8140f4445276cb5160fffe70c86c8d9}} 
\index{Aligner@{Aligner}!Aligner@{Aligner}}
\index{Aligner@{Aligner}!Aligner@{Aligner}}
\subsubsection{\texorpdfstring{Aligner()}{Aligner()}}
{\footnotesize\ttfamily Aligner\+::\+Aligner (\begin{DoxyParamCaption}\item[{\hyperlink{structMatrix}{Matrix}$<$ double $>$}]{sub\+\_\+mat,  }\item[{int}]{gap\+\_\+pen,  }\item[{Gene\+\_\+class}]{gene }\end{DoxyParamCaption})}

Constructor for the \hyperlink{classAligner}{Aligner} class  \+: substitution matrix  \+: sets the gap penalty (the gap penalty is linear)  \+: Gene class of the gene aligned. V gene allows for deletions on the 3\textquotesingle{} side of the genomic template, J gene on the 5\textquotesingle{} , D gene and undefined allow deletion on both sides 

The documentation for this class was generated from the following files\+:\begin{DoxyCompactItemize}
\item 
Aligner.\+h\item 
Aligner.\+cpp\end{DoxyCompactItemize}
