<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>IGoR: ExtractFeatures Class Reference</title>
<link href="../../tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="../../jquery.js"></script>
<script type="text/javascript" src="../../dynsections.js"></script>
<link href="../../search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="../../search/searchdata.js"></script>
<script type="text/javascript" src="../../search/search.js"></script>
<link href="../../doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">IGoR
   </div>
   <div id="projectbrief">InferenceandGenerationofRepertoires</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "../../search",false,'Search');
</script>
<script type="text/javascript" src="../../menudata.js"></script>
<script type="text/javascript" src="../../menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('../../',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="../../d5/d57/classExtractFeatures-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">ExtractFeatures Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Class to extract sequences features (e.g. CDR3) of sequences using alignment and V, J anchors information.  
 <a href="../../d1/d7b/classExtractFeatures.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="../../dc/d5d/ExtractFeatures_8h_source.html">ExtractFeatures.h</a>&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a7b288f4e64f388d34f1e0cd63cd2dcf6"><td class="memItemLeft" align="right" valign="top"><a id="a7b288f4e64f388d34f1e0cd63cd2dcf6"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>ExtractFeatures</b> (const <a class="el" href="../../d1/d7b/classExtractFeatures.html">ExtractFeatures</a> &amp;orig)</td></tr>
<tr class="separator:a7b288f4e64f388d34f1e0cd63cd2dcf6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a70b5cef5ad48234649b114dea3eba942"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#a70b5cef5ad48234649b114dea3eba942">load_VJgenomicTemplates</a> (vector&lt; pair&lt; string, string &gt;&gt; v_genomic, vector&lt; pair&lt; string, string &gt;&gt; j_genomic)</td></tr>
<tr class="memdesc:a70b5cef5ad48234649b114dea3eba942"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to map the templateID and the sequence for V and J.  <a href="#a70b5cef5ad48234649b114dea3eba942">More...</a><br /></td></tr>
<tr class="separator:a70b5cef5ad48234649b114dea3eba942"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a05f401f4f39adb7f33c2b1e3f7f51807"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#a05f401f4f39adb7f33c2b1e3f7f51807">load_VJanchors</a> (string flnV_CDR3_anchors, string flnJ_CDR3_anchors)</td></tr>
<tr class="memdesc:a05f401f4f39adb7f33c2b1e3f7f51807"><td class="mdescLeft">&#160;</td><td class="mdescRight">load data files into GeneFeatures functor class  <a href="#a05f401f4f39adb7f33c2b1e3f7f51807">More...</a><br /></td></tr>
<tr class="separator:a05f401f4f39adb7f33c2b1e3f7f51807"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ace6992bcea2ed1008b5a103291f039c4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#ace6992bcea2ed1008b5a103291f039c4">load_VJanchors</a> (unordered_map&lt; string, size_t &gt; flnV_CDR3_anchors, unordered_map&lt; string, size_t &gt; flnJ_CDR3_anchors)</td></tr>
<tr class="memdesc:ace6992bcea2ed1008b5a103291f039c4"><td class="mdescLeft">&#160;</td><td class="mdescRight">load data files into GeneFeatures functor class  <a href="#ace6992bcea2ed1008b5a103291f039c4">More...</a><br /></td></tr>
<tr class="separator:ace6992bcea2ed1008b5a103291f039c4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac04edfcc22b8b13bdfe24f8786f26df5"><td class="memItemLeft" align="right" valign="top"><a id="ac04edfcc22b8b13bdfe24f8786f26df5"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>print_VgenomicTemplates</b> ()</td></tr>
<tr class="separator:ac04edfcc22b8b13bdfe24f8786f26df5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29cb642660337a3fbaa0446d835ef1fa"><td class="memItemLeft" align="right" valign="top"><a id="a29cb642660337a3fbaa0446d835ef1fa"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>print_JgenomicTemplates</b> ()</td></tr>
<tr class="separator:a29cb642660337a3fbaa0446d835ef1fa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79fcd50fc536ec897d7bae410afa1284"><td class="memItemLeft" align="right" valign="top"><a id="a79fcd50fc536ec897d7bae410afa1284"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>set_sorted_alignments</b> (unordered_map&lt; int, pair&lt; string, unordered_map&lt; Gene_class, vector&lt; <a class="el" href="../../d0/dea/structAlignment__data.html">Alignment_data</a> &gt;&gt;&gt;&gt; *pointer)</td></tr>
<tr class="separator:a79fcd50fc536ec897d7bae410afa1284"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af1dc5d4f1cb7e15227fbafcd2df662d2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="../../db/dda/classCDR3SeqData.html">CDR3SeqData</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#af1dc5d4f1cb7e15227fbafcd2df662d2">extractCDR3</a> (int seq_index)</td></tr>
<tr class="memdesc:af1dc5d4f1cb7e15227fbafcd2df662d2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the CDR3 as an instance of <a class="el" href="../../db/dda/classCDR3SeqData.html" title="Class to store CDR3 information of a sequence. ">CDR3SeqData</a> given the V and J <a class="el" href="../../d0/dea/structAlignment__data.html" title="Stores information on the alignment of one genomic template against the target. ">Alignment_data</a> structs.  <a href="#af1dc5d4f1cb7e15227fbafcd2df662d2">More...</a><br /></td></tr>
<tr class="separator:af1dc5d4f1cb7e15227fbafcd2df662d2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5d1a024949397722223c20d3fc5e6fd4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#a5d1a024949397722223c20d3fc5e6fd4">getVAnchor4Seq</a> (string seq_str, <a class="el" href="../../d0/dea/structAlignment__data.html">Alignment_data</a> v_alig)</td></tr>
<tr class="separator:a5d1a024949397722223c20d3fc5e6fd4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5738a890c09d3cb72ea01693f624b593"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#a5738a890c09d3cb72ea01693f624b593">getJAnchor4Seq</a> (string seq_str, <a class="el" href="../../d0/dea/structAlignment__data.html">Alignment_data</a> j_alig)</td></tr>
<tr class="separator:a5738a890c09d3cb72ea01693f624b593"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa117a4841d8db720b13c0e0b9028579d"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="../../d1/d7b/classExtractFeatures.html#aa117a4841d8db720b13c0e0b9028579d">generateCDR3_csv_line</a> (<a class="el" href="../../db/dda/classCDR3SeqData.html">CDR3SeqData</a> cdr3InputSeq)</td></tr>
<tr class="memdesc:aa117a4841d8db720b13c0e0b9028579d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Function to generate a CDR3 string line to be printed on a file.  <a href="#aa117a4841d8db720b13c0e0b9028579d">More...</a><br /></td></tr>
<tr class="separator:aa117a4841d8db720b13c0e0b9028579d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:aa67318cf2f9b155b533a4ab816fdd8ee"><td class="memItemLeft" align="right" valign="top"><a id="aa67318cf2f9b155b533a4ab816fdd8ee"></a>
unordered_map&lt; string, string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>UMap_v_genomic</b></td></tr>
<tr class="separator:aa67318cf2f9b155b533a4ab816fdd8ee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adf3a835b87f9f274a9b29f830fb7c406"><td class="memItemLeft" align="right" valign="top"><a id="adf3a835b87f9f274a9b29f830fb7c406"></a>
unordered_map&lt; string, string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>UMap_j_genomic</b></td></tr>
<tr class="separator:adf3a835b87f9f274a9b29f830fb7c406"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a11a5252b3e1bd2c12b816dbf3f1cff37"><td class="memItemLeft" align="right" valign="top"><a id="a11a5252b3e1bd2c12b816dbf3f1cff37"></a>
unordered_map&lt; string, size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>UMap_v_CDR3_anchors</b></td></tr>
<tr class="separator:a11a5252b3e1bd2c12b816dbf3f1cff37"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af524e782f4053ae78c0d12374b6ae6f3"><td class="memItemLeft" align="right" valign="top"><a id="af524e782f4053ae78c0d12374b6ae6f3"></a>
unordered_map&lt; string, size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>UMap_j_CDR3_anchors</b></td></tr>
<tr class="separator:af524e782f4053ae78c0d12374b6ae6f3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f90e1dce7c28bc411cebfd72cd63257"><td class="memItemLeft" align="right" valign="top"><a id="a8f90e1dce7c28bc411cebfd72cd63257"></a>
unordered_map&lt; int, pair&lt; string, unordered_map&lt; Gene_class, vector&lt; <a class="el" href="../../d0/dea/structAlignment__data.html">Alignment_data</a> &gt; &gt; &gt; &gt; *&#160;</td><td class="memItemRight" valign="bottom"><b>p_sorted_alignments</b></td></tr>
<tr class="separator:a8f90e1dce7c28bc411cebfd72cd63257"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Class to extract sequences features (e.g. CDR3) of sequences using alignment and V, J anchors information. </p>
<dl class="section author"><dt>Author</dt><dd>C. Olivares </dd></dl>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a id="af1dc5d4f1cb7e15227fbafcd2df662d2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af1dc5d4f1cb7e15227fbafcd2df662d2">&#9670;&nbsp;</a></span>extractCDR3()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="../../db/dda/classCDR3SeqData.html">CDR3SeqData</a> ExtractFeatures::extractCDR3 </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>seq_index</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Get the CDR3 as an instance of <a class="el" href="../../db/dda/classCDR3SeqData.html" title="Class to store CDR3 information of a sequence. ">CDR3SeqData</a> given the V and J <a class="el" href="../../d0/dea/structAlignment__data.html" title="Stores information on the alignment of one genomic template against the target. ">Alignment_data</a> structs. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq_index</td><td></td></tr>
    <tr><td class="paramname">V_alignment</td><td></td></tr>
    <tr><td class="paramname">J_alignment</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd><a class="el" href="../../db/dda/classCDR3SeqData.html" title="Class to store CDR3 information of a sequence. ">CDR3SeqData</a> </dd></dl>

</div>
</div>
<a id="aa117a4841d8db720b13c0e0b9028579d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa117a4841d8db720b13c0e0b9028579d">&#9670;&nbsp;</a></span>generateCDR3_csv_line()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string ExtractFeatures::generateCDR3_csv_line </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="../../db/dda/classCDR3SeqData.html">CDR3SeqData</a>&#160;</td>
          <td class="paramname"><em>cdr3InputSeq</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Function to generate a CDR3 string line to be printed on a file. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">cdr3InputSeq</td><td>is <a class="el" href="../../db/dda/classCDR3SeqData.html" title="Class to store CDR3 information of a sequence. ">CDR3SeqData</a> instance </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>csvline line string to be printed on a file. </dd></dl>

</div>
</div>
<a id="a5738a890c09d3cb72ea01693f624b593"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5738a890c09d3cb72ea01693f624b593">&#9670;&nbsp;</a></span>getJAnchor4Seq()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int ExtractFeatures::getJAnchor4Seq </td>
          <td>(</td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>seq_str</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="../../d0/dea/structAlignment__data.html">Alignment_data</a>&#160;</td>
          <td class="paramname"><em>j_alig</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Calculate the J anchor in reference to the input sequence </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq_str</td><td></td></tr>
    <tr><td class="paramname">j_alig</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>cdr3_j_read_anch the anchor in reference to the input sequence. </dd></dl>

</div>
</div>
<a id="a5d1a024949397722223c20d3fc5e6fd4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5d1a024949397722223c20d3fc5e6fd4">&#9670;&nbsp;</a></span>getVAnchor4Seq()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int ExtractFeatures::getVAnchor4Seq </td>
          <td>(</td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>seq_str</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="../../d0/dea/structAlignment__data.html">Alignment_data</a>&#160;</td>
          <td class="paramname"><em>v_alig</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Calculate the V anchor in reference to the input sequence </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">seq_str</td><td></td></tr>
    <tr><td class="paramname">v_alig</td><td></td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>cdr3_v_read_anch the anchor in reference to the input sequence. </dd></dl>

</div>
</div>
<a id="a05f401f4f39adb7f33c2b1e3f7f51807"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a05f401f4f39adb7f33c2b1e3f7f51807">&#9670;&nbsp;</a></span>load_VJanchors() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void ExtractFeatures::load_VJanchors </td>
          <td>(</td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>flnV_CDR3_anchors</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>flnJ_CDR3_anchors</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load data files into GeneFeatures functor class </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">flnV_CDR3_anchors</td><td>CDR3 anchors filename for V genes </td></tr>
    <tr><td class="paramname">flnJ_CDR3_anchors</td><td>CDR3 anchors filename for J genes </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ace6992bcea2ed1008b5a103291f039c4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ace6992bcea2ed1008b5a103291f039c4">&#9670;&nbsp;</a></span>load_VJanchors() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void ExtractFeatures::load_VJanchors </td>
          <td>(</td>
          <td class="paramtype">unordered_map&lt; string, size_t &gt;&#160;</td>
          <td class="paramname"><em>v_CDR3_anchors</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unordered_map&lt; string, size_t &gt;&#160;</td>
          <td class="paramname"><em>j_CDR3_anchors</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>load data files into GeneFeatures functor class </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">v_CDR3_anchors</td><td>unordered_map of sequence description and position for V genes CDR3 anchors. </td></tr>
    <tr><td class="paramname">j_CDR3_anchors</td><td>unordered_map of sequence description and position for J genes CDR3 anchors. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="a70b5cef5ad48234649b114dea3eba942"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a70b5cef5ad48234649b114dea3eba942">&#9670;&nbsp;</a></span>load_VJgenomicTemplates()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void ExtractFeatures::load_VJgenomicTemplates </td>
          <td>(</td>
          <td class="paramtype">vector&lt; pair&lt; string, string &gt;&gt;&#160;</td>
          <td class="paramname"><em>v_genomic</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; pair&lt; string, string &gt;&gt;&#160;</td>
          <td class="paramname"><em>j_genomic</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Function to map the templateID and the sequence for V and J. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">v_genomic</td><td>vector with the genomic templates for V genes </td></tr>
    <tr><td class="paramname">j_genomic</td><td>vector with the genomic templates for J genes </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd></dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li><a class="el" href="../../dc/d5d/ExtractFeatures_8h_source.html">ExtractFeatures.h</a></li>
<li>ExtractFeatures.cpp</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="../../doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
