\section{Class List}
Here are the classes, structs, unions and interfaces with brief descriptions\+:\begin{DoxyCompactList}
\item\contentsline{section}{\hyperlink{structAdjacency__list}{Adjacency\+\_\+list} \\*I\+GoR\textquotesingle{}s Bayesian Network adjacency list }{\pageref{structAdjacency__list}}{}
\item\contentsline{section}{\hyperlink{classAligner}{Aligner} \\*A modified Smith-\/\+Waterman alignment class }{\pageref{classAligner}}{}
\item\contentsline{section}{\hyperlink{structAlignment__data}{Alignment\+\_\+data} \\*Stores information on the alignment of one genomic template against the target }{\pageref{structAlignment__data}}{}
\item\contentsline{section}{\hyperlink{classBest__scenarios__counter}{Best\+\_\+scenarios\+\_\+counter} \\*Records the N best scenarios realizations and mismatches }{\pageref{classBest__scenarios__counter}}{}
\item\contentsline{section}{\hyperlink{classCDR3SeqData}{C\+D\+R3\+Seq\+Data} \\*Class to store C\+D\+R3 information of a sequence }{\pageref{classCDR3SeqData}}{}
\item\contentsline{section}{\hyperlink{classCounter}{Counter} \\*Scenario statistics recording abstract class }{\pageref{classCounter}}{}
\item\contentsline{section}{\hyperlink{classCoverage__err__counter}{Coverage\+\_\+err\+\_\+counter} \\*Records the number of time each genomic position is observed with or without error }{\pageref{classCoverage__err__counter}}{}
\item\contentsline{section}{\hyperlink{structD__position__comparator}{D\+\_\+position\+\_\+comparator} }{\pageref{structD__position__comparator}}{}
\item\contentsline{section}{\hyperlink{classDeletion}{Deletion} \\*\hyperlink{classDeletion}{Deletion} recombination event }{\pageref{classDeletion}}{}
\item\contentsline{section}{\hyperlink{classDinucl__markov}{Dinucl\+\_\+markov} \\*Dinucleotide insertion Markov model }{\pageref{classDinucl__markov}}{}
\item\contentsline{section}{\hyperlink{classEnum__fast__memory__dual__key__map}{Enum\+\_\+fast\+\_\+memory\+\_\+dual\+\_\+key\+\_\+map$<$ K1, K2, V $>$} }{\pageref{classEnum__fast__memory__dual__key__map}}{}
\item\contentsline{section}{\hyperlink{classEnum__fast__memory__map}{Enum\+\_\+fast\+\_\+memory\+\_\+map$<$ K, V $>$} }{\pageref{classEnum__fast__memory__map}}{}
\item\contentsline{section}{\hyperlink{classError__rate}{Error\+\_\+rate} \\*Abstract class for generic error models behavior }{\pageref{classError__rate}}{}
\item\contentsline{section}{\hyperlink{classErrors__counter}{Errors\+\_\+counter} \\*\hyperlink{classCounter}{Counter} recording the number of genomic nucleotides and errors/mismatch per scenario }{\pageref{classErrors__counter}}{}
\item\contentsline{section}{\hyperlink{structEvent__comparator}{Event\+\_\+comparator} }{\pageref{structEvent__comparator}}{}
\item\contentsline{section}{\hyperlink{structEvent__realization}{Event\+\_\+realization} \\*Unit that stores an event realization name, value and index }{\pageref{structEvent__realization}}{}
\item\contentsline{section}{\hyperlink{classExtractFeatures}{Extract\+Features} \\*Class to extract sequences features (e.\+g. C\+D\+R3) of sequences using alignment and V, J anchors information }{\pageref{classExtractFeatures}}{}
\item\contentsline{section}{\hyperlink{structgen__CDR3__data}{gen\+\_\+\+C\+D\+R3\+\_\+data} }{\pageref{structgen__CDR3__data}}{}
\item\contentsline{section}{\hyperlink{classGene__choice}{Gene\+\_\+choice} \\*Gene\+Choice recombination event }{\pageref{classGene__choice}}{}
\item\contentsline{section}{\hyperlink{classGenModel}{Gen\+Model} \\*High level V(\+D)J generative model }{\pageref{classGenModel}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01Event__safety_01_4}{std\+::hash$<$ Event\+\_\+safety $>$} }{\pageref{structstd_1_1hash_3_01Event__safety_01_4}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01Gene__class_01_4}{std\+::hash$<$ Gene\+\_\+class $>$} }{\pageref{structstd_1_1hash_3_01Gene__class_01_4}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01Int__Str_01_4}{std\+::hash$<$ Int\+\_\+\+Str $>$} }{\pageref{structstd_1_1hash_3_01Int__Str_01_4}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01Seq__type_01_4}{std\+::hash$<$ Seq\+\_\+type $>$} }{\pageref{structstd_1_1hash_3_01Seq__type_01_4}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01std_1_1pair_3_01Gene__class_00_01Seq__side_01_4_01_4}{std\+::hash$<$ std\+::pair$<$ Gene\+\_\+class, Seq\+\_\+side $>$ $>$} }{\pageref{structstd_1_1hash_3_01std_1_1pair_3_01Gene__class_00_01Seq__side_01_4_01_4}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01std_1_1pair_3_01Seq__type_00_01Seq__side_01_4_01_4}{std\+::hash$<$ std\+::pair$<$ Seq\+\_\+type, Seq\+\_\+side $>$ $>$} }{\pageref{structstd_1_1hash_3_01std_1_1pair_3_01Seq__type_00_01Seq__side_01_4_01_4}}{}
\item\contentsline{section}{\hyperlink{structstd_1_1hash_3_01std_1_1tuple_3_01Event__type_00_01Gene__class_00_01Seq__side_01_4_01_4}{std\+::hash$<$ std\+::tuple$<$ Event\+\_\+type, Gene\+\_\+class, Seq\+\_\+side $>$ $>$} }{\pageref{structstd_1_1hash_3_01std_1_1tuple_3_01Event__type_00_01Gene__class_00_01Seq__side_01_4_01_4}}{}
\item\contentsline{section}{\hyperlink{classHypermutation__full__Nmer__errorrate}{Hypermutation\+\_\+full\+\_\+\+Nmer\+\_\+errorrate} \\*A non-\/additive context dependent hypermutation model }{\pageref{classHypermutation__full__Nmer__errorrate}}{}
\item\contentsline{section}{\hyperlink{classHypermutation__global__errorrate}{Hypermutation\+\_\+global\+\_\+errorrate} \\*An additive (independent site) context dependent hypermutation model }{\pageref{classHypermutation__global__errorrate}}{}
\item\contentsline{section}{\hyperlink{classInsertion}{Insertion} \\*\hyperlink{classInsertion}{Insertion} recombination events }{\pageref{classInsertion}}{}
\item\contentsline{section}{\hyperlink{classInt__Str}{Int\+\_\+\+Str} }{\pageref{classInt__Str}}{}
\item\contentsline{section}{\hyperlink{structinverse__offset__comparator}{inverse\+\_\+offset\+\_\+comparator} }{\pageref{structinverse__offset__comparator}}{}
\item\contentsline{section}{\hyperlink{structMatrix}{Matrix$<$ T $>$} }{\pageref{structMatrix}}{}
\item\contentsline{section}{\hyperlink{classModel__marginals}{Model\+\_\+marginals} \\*Encapsulates the marginal probabilities/posterior frequency for each recombination event\textquotesingle{}s realization }{\pageref{classModel__marginals}}{}
\item\contentsline{section}{\hyperlink{classModel__Parms}{Model\+\_\+\+Parms} \\*Implements I\+GoR\textquotesingle{}s Bayesian Network structure }{\pageref{classModel__Parms}}{}
\item\contentsline{section}{\hyperlink{structnull__delete}{null\+\_\+delete$<$ T $>$} \\*Declare a \hyperlink{structnull__delete}{null\+\_\+delete} function }{\pageref{structnull__delete}}{}
\item\contentsline{section}{\hyperlink{structoffset__comp}{offset\+\_\+comp} }{\pageref{structoffset__comp}}{}
\item\contentsline{section}{\hyperlink{classPgen__counter}{Pgen\+\_\+counter} \\*Estimates sequences generation probability }{\pageref{classPgen__counter}}{}
\item\contentsline{section}{\hyperlink{classRec__Event}{Rec\+\_\+\+Event} \\*Recombination event class (I\+GoR\textquotesingle{}s graph nodes) }{\pageref{classRec__Event}}{}
\item\contentsline{section}{\hyperlink{classSingle__error__rate}{Single\+\_\+error\+\_\+rate} \\*Independent single nucleotide error model }{\pageref{classSingle__error__rate}}{}
\end{DoxyCompactList}
